### ISSUE TITLE
Functionality to "forgive" null values 

### ISSUE THREAD

**Issue by Kimonili** [astropenguin/pandas-dataclasses#115]
Hi there,

I was wondering whether there is a way to ignore `null` values from columns with a specified type hint, say `str` if majority of observations are of the correct type? It currently throws an error if that happens. 

Thank you for developing this package!

**Comment by astropenguin** [comment]
@Kimonili Thank you for the question! Just to confirm, do you mean you want to handle data like `["a", "b", None, "c", ...]` which the current type hint like `Data[str]` cannot handle properly?

**Comment by Kimonili** [comment]
Thanks, @astropenguin, exactly. I believe that the current functionality will throw an error if it comes across a None value in a string column as you have mentioned in your example. Is there a workaround for something like this?

**Comment by astropenguin** [comment]
@Kimonili Currently we have two options. If you do not need type casting by pandas-dataclasses (i.e. input data already have correct type), you can simply use `Any` type:
```python
from typing import Any


@dataclass
class Sample:
    data: Data[Any]
```
Because `Data[Any]` is equivalent to `typing.Collection[Any]`, it accepts collections (e.g. list, tuple, ...) of any type.

If you need type casting, you can use `Union` type to add an alternative type hint:
```python
from typing import Any, Collection, Union


@dataclass
class Sample:
    data: Union[Data[str], Collection[Union[str, None]]]
```
In this case, `Collection[Union[str, None]]` accepts string data with missing values. In Python 3.10 and later, it is simpler to write:
```python
@dataclass
class Sample:
    data: Data[str] | Collection[str | None]
```

**Comment by Kimonili** [comment]
Hey @astropenguin, thank you for the speedy reply!

I am primarily interested in the second solution, as I'd really like to take advantage of the stability of the pandas-dataclasses structure compared to regular dataframes. However, I've done some tests, and the second solution fails for `int` type columns (it works well for strings, floats, and boolean types columns). 

See example below to reproduce the above:

```python
@dataclass
class Weather(AsDataFrame):
    """Weather information."""

    str_col: Union[Data[str], Collection[Union[str, None]]]
    int_col: Union[Data[int], Collection[Union[int, None]]]
    float_col: Union[Data[float], Collection[Union[float, None]]]
    bool_col: Union[Data[bool], Collection[Union[bool, None]]]


df = Weather.new(
    ['2020', None, '2021', '2021', '2022'],
    [1,2,3, None,5],
    [7.1, 24.3, 5.4, None, 4.9],
    [True, False, None, False, False]
)
```

Any idea why `int` type columns do not behave in the same way as the other data types?

**Comment by astropenguin** [comment]
@Kimonili Thank you for the details! My apologies, I thought the error came from a static type checker, not the runtime code... In this case, please use pandas-specific data type for integers like `Data[pandas.Int64Dtype]`. The error happens because integer type generally does not define null (or NaN). With `Data[int]`, pandas-dataclasses internally runs a code similar to the following:
```python
pandas.array([1, 2, 3, None, 5], dtype=int)
```
This will also raise the same error you reported. In contrast, pandas-specific data types accept null values, so you can avoid the error. The workarounds I proposed above are only for static type checks. If you do not mind them, specifying `Data[pandas.Int64Dtype]` is just enough.

> it works well for strings, floats, and boolean types columns

I think it might be better to use them (e.g. `StringDtype`, `Float64Dtype`, `BooleanDtype`) for these columns as well. For example, `None` will be unintentionally converted to `False` in the case of `Data[bool]`. With `Data[pandas.BooleanDtype]`, you can keep distinguishing between `False` and `None`.

**Comment by Kimonili** [comment]
Hey @astropenguin,

After following your suggestions and run the tests for all data types, I got the following errors:

- For numpy arrays inputs:

```python
@dataclass
class Weather(AsDataFrame):
    """Weather information."""

    str_col: Data[StringDtype]
    int_col: Data[Int64Dtype]
    float_col: Data[Float64Dtype]
    bool_col: Data[BooleanDtype]


df = Weather.new(
    np.array(['2020', 'None', '2021', '2021', '2022']),
    np.array([1,2,3,3,5]),
    np.array([7.1, 24.3, 5.4, 3.2, 4.9]),
    np.array([True, False, False, False, False])
```
I got this error:

```python
Output exceeds the [size limit](command:workbench.action.openSettings?[). Open the full output data [in a text editor](command:workbench.action.openLargeOutput?b52789ef-1cbb-474f-bb5d-fdfbef2d93d1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In [87], line 13
      9     float_col: Data[Float64Dtype]
     10     bool_col: Data[BooleanDtype]
---> 13 df = Weather.new(
     14     np.array(['2020', 'None', '2021', '2021', '2022']),
     15     np.array([1,2,3,3,5]),
     16     np.array([7.1, 24.3, 5.4, 3.2, 4.9]),
     17     np.array([True, False, False, False, False])
     18 )

File ~/Repositories/rx-item-recommendations/.venv/lib/python3.9/site-packages/pandas_dataclasses/core/mixins.py:64, in As.new.<locals>.new(cls, *args, **kwargs)
     61 @wraps(init)
     62 def new(cls: Any, *args: Any, **kwargs: Any) -> Any:
     63     if issubclass(cls.__pandas_factory__, pd.DataFrame):
---> 64         return asdataframe(cls(*args, **kwargs))
     65     elif issubclass(cls.__pandas_factory__, pd.Series):
     66         return asseries(cls(*args, **kwargs))

File ~/Repositories/rx-item-recommendations/.venv/lib/python3.9/site-packages/pandas_dataclasses/core/parsers.py:43, in asdataframe(obj, factory)
     41 spec = Spec.from_dataclass(type(obj)) @ obj
     42 attrs = get_attrs(spec)
---> 43 data = get_data(spec)
     44 index = get_index(spec)
...
--> 326     cls = cast(ExtensionDtype, dtype).construct_array_type()
    327     return cls._from_sequence(data, dtype=dtype, copy=copy)
    329 if dtype is None:

TypeError: construct_array_type() missing 1 required positional argument: 'self'
```

- For list inputs:
```python
@dataclass
class Weather(AsDataFrame):
    """Weather information."""

    str_col: Data[StringDtype]
    int_col: Data[Int64Dtype]
    float_col: Data[Float64Dtype]
    bool_col: Data[BooleanDtype]


df = Weather.new(
    ['2020', None, '2021', '2021', '2022'],
    [1,2,3,3,5],
    [7.1, 24.3, 5.4, 3.2, 4.9],
    [True, False, False, False, False]
)
```
I got the same error as above + this hint from pylance for each value of each column:

```python
Argument of type "list[str | None]" cannot be assigned to parameter "str_col" of type "Data[StringDtype]"
  "Literal['2020']" is incompatible with "StringDtype"
  Type "None" cannot be assigned to type "StringDtype"
  "Literal['2021']" is incompatible with "StringDtype"
  "Literal['2021']" is incompatible with "StringDtype"
  "Literal['2022']" is incompatible with "StringDtype"Pylance[reportGeneralTypeIssues](https://github.com/microsoft/pylance-release/blob/main/DIAGNOSTIC_SEVERITY_RULES.md#diagnostic-severity-rules)
```

**Comment by astropenguin** [comment]
Hi @Kimonili, I guess you are using older pandas-dataclasses...? With the latest version (v0.9.0), we can run both codes (here I made [an example Google Colab notebook][URL]). I am sorry for the inconvenience, but could you consider updating the package? For the last error by Pylance, you still need to use Union type:
```python
str_col: Union[Data[StringDtype], Sequence[Union[str, None]]]
```

[URL]: https://colab.research.google.com/drive/1SKUfYkW3wrdhsH8Us3Wo-9IgJDN3ceYH

**Comment by Kimonili** [comment]
Hey @astropenguin, thank you so much! It now has the expected behaviour 😃

**Comment by astropenguin** [comment]
Hi @Kimonili, I released v0.10.2. Now you can use the following type specification:

```python
str_col: Data[Union[StringDtype, str, None]]  # equivalent to Collection[StringDtype | str | None]
```
